File: payment_routes.js
// payment_routes.js
import express from "express";
import fetch from "node-fetch";
import { db, admin } from "./firebase.js"; // Firebase Admin SDK

const router = express.Router();

/* ------------------ PAYMONGO CONFIG ------------------ */
const PAYMONGO_SECRET_KEY = process.env.PAYMONGO_SECRET_KEY || "sk_test_uYAyatPB8sNrDkLispMVrLh4";

const authHeader = {
  Authorization: "Basic " + Buffer.from(PAYMONGO_SECRET_KEY + ":").toString("base64"),
  "Content-Type": "application/json",
};

/* ------------------ HELPERS ------------------ */
async function pmPost(url, body) {
  const res = await fetch(url, {
    method: "POST",
    headers: authHeader,
    body: JSON.stringify(body),
  });

  const data = await res.json();

  if (!res.ok) {
    const msg = data?.errors?.[0]?.detail || JSON.stringify(data);
    throw new Error(`PayMongo error: ${msg}`);
  }

  return data;
}

/* ------------------ ROUTES ------------------ */

/**
 * @route POST /paymongo/:type/intent
 * @desc Create payment intent for GCash or GrabPay
 * @body { amount, billing }
 */
router.post("/:type/intent", async (req, res) => {
  try {
    const paymentType = req.params.type.toLowerCase();
    const { amount, billing } = req.body || {};

    if (!amount || isNaN(amount)) {
      return res.status(400).json({ error: "Invalid amount" });
    }

    const amountCentavos = Math.round(Number(amount) * 100);

    // Prepare billing info
    const billingData = {
      name: billing?.name || `${paymentType.toUpperCase()} Payer`,
      email: billing?.email || "payer@example.com",
    };

    // Add phone depending on payment type
    if (paymentType === "gcash") billingData.gcashNumber = billing?.phone || "09123456789";
    if (paymentType === "grab_pay") billingData.phone = billing?.phone || "09123456789";

    const source = await pmPost("https://api.paymongo.com/v1/sources", {
      data: {
        attributes: {
          amount: amountCentavos,
          redirect: {
            success: "https://iskolardev.online/payment-success",
            failed: "https://iskolardev.online/payment-failed",
          },
          type: paymentType, // "gcash" or "grabpay"
          currency: "PHP",
          billing: billingData,
        },
      },
    });

    const checkoutUrl = source?.data?.attributes?.redirect?.checkout_url;
    const sourceId = source?.data?.id;

    if (!checkoutUrl) {
      return res.status(500).json({ error: "Failed to create checkout URL" });
    }

    // Save payment log to Firestore
    await db.collection("payments").doc(sourceId).set({
      amount,
      billing,
      paymentType,
      sourceId,
      status: "pending",
      createdAt: admin.firestore.FieldValue.serverTimestamp(),
    });

    res.json({ checkoutUrl, sourceId });
  } catch (err) {
    console.error(`❌ Error in /paymongo/${req.params.type}/intent:`, err);
    res.status(500).json({ error: err.message || "Server error" });
  }
});

/**
 * @route POST /paymongo/:type/webhook
 * @desc Webhook handler for payment updates
 */
router.post("/:type/webhook", express.json({ type: "*/*" }), async (req, res) => {
  try {
    const paymentId = req.body?.data?.id;
    const event = req.body?.data?.attributes?.type;

    if (!paymentId || !event) return res.sendStatus(400);

    let update = {};
    if (event === "payment.paid") {
      update = { status: "paid", paidAt: admin.firestore.FieldValue.serverTimestamp() };
    } else if (event === "payment.failed") {
      update = { status: "failed", failedAt: admin.firestore.FieldValue.serverTimestamp() };
    }

    if (Object.keys(update).length) {
      await db.collection("payments").doc(paymentId).set(update, { merge: true });
    }

    res.sendStatus(200);
  } catch (err) {
    console.error(`❌ Webhook error for ${req.params.type}:`, err);
    res.sendStatus(500);
  }
});

export default router;

File: payment_routes.py
# routes/payment_routes.py
from flask import Blueprint, request, jsonify
import os
import requests
from app import db
from firebase_admin import firestore
from base64 import b64encode
from datetime import datetime

payment_bp = Blueprint("payment", __name__)

# ------------------ PAYMONGO CONFIG ------------------
PAYMONGO_SECRET_KEY = os.environ.get("PAYMONGO_SECRET_KEY", "sk_test_uYAyatPB8sNrDkLispMVrLh4")
AUTH_HEADER = {
    "Authorization": "Basic " + b64encode(f"{PAYMONGO_SECRET_KEY}:".encode()).decode(),
    "Content-Type": "application/json"
}

# ------------------ HELPER ------------------
def pm_post(url, body):
    res = requests.post(url, headers=AUTH_HEADER, json=body)
    data = res.json()
    if not res.ok:
        msg = data.get("errors", [{}])[0].get("detail", str(data))
        raise Exception(f"PayMongo error: {msg}")
    return data

# ------------------ ROUTES ------------------

@payment_bp.route("/<string:payment_type>/intent", methods=["POST"])
def create_payment_intent(payment_type):
    try:
        payment_type = payment_type.lower()
        body = request.json or {}
        amount = body.get("amount")
        billing = body.get("billing", {})

        if not amount or not str(amount).replace(".", "", 1).isdigit():
            return jsonify({"error": "Invalid amount"}), 400

        amount_centavos = round(float(amount) * 100)

        billing_data = {
            "name": billing.get("name", f"{payment_type.upper()} Payer"),
            "email": billing.get("email", "payer@example.com")
        }

        if payment_type == "gcash":
            billing_data["gcashNumber"] = billing.get("phone", "09123456789")
        if payment_type == "grab_pay":
            billing_data["phone"] = billing.get("phone", "09123456789")

        source = pm_post("https://api.paymongo.com/v1/sources", {
            "data": {
                "attributes": {
                    "amount": amount_centavos,
                    "redirect": {
                        "success": "https://iskolardev.online/payment-success",
                        "failed": "https://iskolardev.online/payment-failed"
                    },
                    "type": payment_type,
                    "currency": "PHP",
                    "billing": billing_data
                }
            }
        })

        checkout_url = source.get("data", {}).get("attributes", {}).get("redirect", {}).get("checkout_url")
        source_id = source.get("data", {}).get("id")

        if not checkout_url:
            return jsonify({"error": "Failed to create checkout URL"}), 500

        # Save payment log to Firestore
        db.collection("payments").document(source_id).set({
            "amount": amount,
            "billing": billing,
            "paymentType": payment_type,
            "sourceId": source_id,
            "status": "pending",
            "createdAt": firestore.SERVER_TIMESTAMP
        })

        return jsonify({"checkoutUrl": checkout_url, "sourceId": source_id})

    except Exception as e:
        print(f"❌ Error in /paymongo/{payment_type}/intent:", e)
        return jsonify({"error": str(e) or "Server error"}), 500


@payment_bp.route("/<string:payment_type>/webhook", methods=["POST"])
def webhook(payment_type):
    try:
        body = request.json or {}
        payment_id = body.get("data", {}).get("id")
        event = body.get("data", {}).get("attributes", {}).get("type")

        if not payment_id or not event:
            return "", 400

        update = {}
        if event == "payment.paid":
            update = {"status": "paid", "paidAt": firestore.SERVER_TIMESTAMP}
        elif event == "payment.failed":
            update = {"status": "failed", "failedAt": firestore.SERVER_TIMESTAMP}

        if update:
            db.collection("payments").document(payment_id).set(update, merge=True)

        return "", 200

    except Exception as e:
        print(f"❌ Webhook error for {payment_type}:", e)
        return "", 500
